### ISSUE TITLE
Implementation seems to differ from paper

### ISSUE THREAD

**Issue by ltrottier** [Kautenja/keras-pyramid-pooling-module#1]
Hi,

Thanks for uploading your code.

I was comparing it with the description of the module in the original paper, and I saw a possible difference: bin_size seems different than pool_size.

The pool_size of K.pool_2d is using the bin_size (see https://github.com/Kautenja/keras-pyramid-pooling-module/blob/master/keras_pyramid_pooling_module/pyramid_pooling_module.py#L209). But in the paper, the bin_size seems to be the size of the feature map after pooling (p.4 of paper, we can read: "The coarsest level highlighted in red is global pooling to generate a single bin output."). This suggests that you should use something like this instead:
```python
[...]
if self.data_format == 'channels_last':
    input_shape = np.array(K.int_shape(input_)[1:-1])
if self.data_format == 'channels_first':
    input_shape = np.array(K.int_shape(input_)[2:])
[...]
for (level, bin_size) in enumerate(self.bin_sizes):
    pool_size = (input_shape / bin_size).astype('int')
  
  if np.any(pool_size * bin_size != input_shape):
        msg = (
            f'Input shape is not a multiple of bin_size:\n'
            f'    Input shape is: {input_shape}\n'
            f'    Current bin_size is: {bin_size}\n'
            f'    bin_sizes: {self.bin_sizes}'
        )
        raise ValueError(msg)

    x = K.pool2d(
        input_,
        pool_size,
        strides=(bin_size, bin_size),
        padding=self.pool_padding,
        pool_mode=self.pool_mode,
        data_format=self.data_format,
    )

```
which is similar to spatial pyramid pooling.  What do you think?

**Comment by ltrottier** [comment]
"The idea of sub-region average pooling actually is quite similar to Spatial Pyramid Pooling in SPPNet."

https://towardsdatascience.com/review-pspnet-winner-in-ilsvrc-2016-semantic-segmentation-scene-parsing-e089e5df177d

**Comment by Kautenja** [comment]
Hmm, it's been a while since I worked with this code or model. I'll give the paper a re-read this week and review difference here. Thanks for catching this, have you compared outputs between the two implementations by chance?

**Comment by ltrottier** [comment]
No problem.

The outputs are different, since the pool shape are not the same.

**Comment by Kautenja** [comment]
I see your point now. My original interpretation was essentially inverted in that the layer is doing 1x1, 2x2, etc. pooling opposed to _producing_ a 1x1, 2x2, etc. matrix. That would probably explain why my PSPNet implementation failed to converge when I was messing with this a few months back lol. I'll get this injected and deployed shortly! Thanks for catching the bug :)

**Comment by Kautenja** [comment]
Deployed! I decided to allow arbitrary input shapes though since the up-sampling can resize to arbitrary size.

**Comment by ltrottier** [comment]
You're welcome!

**Comment by daniel-pp** [comment]
Shouldn't the strides parameter of K.pool2d also be updated accordingly, i.e. `strides=tuple(dim // bin_size for dim in output_shape)` instead of `strides=(bin_size, bin_size)`?

**Comment by Kautenja** [comment]
Yea you're right, the strides parameter should be equal to the pool size. I double checked the output of the pooling layers as the bin size increases to validate the claim.

before: the image size decreases as the bin size increases -- incorrect

```
(?, 1, 1, 3)
(?, 76, 114, 3)
(?, 67, 101, 3)
(?, 42, 63, 3)
```

after suggested change: image size is strictly monotonically increasing as bin size increases -- correct

```
(?, 1, 1, 3)
(?, 151, 227, 3)
(?, 201, 302, 3)
(?, 251, 377, 3)
```

example.ipynb has some code to check the outputs of the layer qualitatively if anyone is curious.

**Comment by daniel-pp** [comment]
In the update you have also changed the number of output filters of the convolution to just 1. However, the [https://github.com/hszhao/PSPNet](official implementation) speaks for the previous version, see e.g. [https://github.com/hszhao/PSPNet/blob/4b53f1c97a5921a99a965a60c0940eec2d46bb06/evaluation/prototxt/pspnet101_cityscapes_713.prototxt#L6733](this line) where the 1x1 convolution in the pyramid pooling module has 512 output filters.

**Comment by daniel-pp** [comment]
And one more suggestion would be to include (the possibility) to add batch normalization after the convolution as was also done in the official implementation.

In any case, thank you for the nice code!

**Comment by Kautenja** [comment]
wow the paper is practically useless. They don't mention anything about the number of filters in the module and at one point make reference to tuning BN parameters, but describe it nowhere in the architecture. Very poor scientific reporting.

**Comment by Kautenja** [comment]
I'm not seeing any BN after convolution in the official source file [here](https://github.com/hszhao/PSPNet/blob/master/src/caffe/layers/spp_layer.cpp). I also see no reference to the 512 value for feature maps. Can you provide a link to back up those claims?

**Comment by Kautenja** [comment]
What i'll do is parameterize the number of filters like in a standard Conv2D layer, but default them to 512. It also looks like I need to disable the bias terms by default and introduce optional BN like you mentioned.

**Comment by Kautenja** [comment]
the number of filters has been updated and I've opened an issue for the batch normalization in case I don't have time to get to it today (#3).
